src/control/family.lean

```diff
@@ -0,0 +1,666 @@
+/-
+Copyright (c) 2019 Simon Hudon. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Simon Hudon
+-/
+import category_theory.category
+import category_theory.types
+import logic.relation
+
+/-!
+Indexed type families and their categorical structure.
+
+Features:
+
+`fam I`      : family of types and its instance of `category`
+
+Also, support functions for operating with n-tuples of types, such as:
+
+`append1 α β`    : append type J-indexed `β` family to I-indexed `α` family to obtain a `(I⊕J)`-indexed family
+`drop α`         : drops the last element of an `(I⊕J)`-indexed family
+`last α`         : returns the last element of an `(I⊕J)`-indexed family
+`append_fun f g` : appends two families of functions `f` and `g`
+`drop_fun f`     : drops the last function family from an `(I⊕J)`-indexed family
+`last_fun f`     : returns the last function family of a `(I⊕J)`-indexed family
+
+Since e.g. `append1 α.drop α.last` is propositionally equal to `α` but not definitionally equal
+to it, we need support functions and lemmas to mediate between constructions.
+-/
+
+universes u v w
+
+open category_theory
+
+/-- Type family indexed by `I`; we call `fam (I ⊕ J)` the product
+of `fam I` and `fam J` (despite being formulated using a sum type) -/
+@[reducible]
+def fam (I : Type u) := I → Type u
+
+instance {I} : has_one (fam I) :=
+⟨ λ _, punit ⟩
+
+namespace fam
+
+variables  {I : Type u}
+
+/-- retain the left side of a product family -/
+def drop {α : Type u} : fam (I ⊕ α) → fam I :=
```

Comment:
The use of alpha in the next three lemmas is really confusing, since it should be J. Everywhere else, the indices are J, and alpha is the indexed family; only here is alpha the index. It is even more confusing since the documentation in the header also uses alpha for `drop`, but at the family, not the indices.